lib/Hakyll/Core/Runtime.hs

```diff
@@ -321,13 +323,19 @@ chase id' = do
                 let depDone =
                         depId `S.member` done ||
                         (depId, depSnapshot) `S.member` snapshots
+                    actualDep = [(depId, depSnapshot) | not depDone]
+
+                -- Update the dependencies using the current snapshot as context
+                -- This ensures that dependency cycles on snapshots are caught as well
+                -- See Issue #878.
+                unless depDone $
+                    modifyRuntimeState $ \s -> s {
+                        runtimeDependencies = M.insertWith S.union (id', depSnapshot) (S.fromList actualDep) (runtimeDependencies s) }
```

Comment:
`depSnapshot` is unrelated to `id'`, so `(id', depSnapshot)` is not really a node in the dependency graph. And this is not just a theoretical concern, here's a test:

```haskell
    logger  <- Logger.new Logger.Error
    (ec, _) <- run RunModeNormal testConfiguration logger $ do

        create ["one.html"] $ do
            route idRoute
            compile $ do
                {- one:1 -} makeItem ("one-one" :: String) >>= saveSnapshot "one"
                {- one:2 -} _ <- (loadSnapshotBody "two.html" "two") :: Compiler String
                {- one:3 -} makeItem ("one-three" :: String) >>= saveSnapshot "three"
                {- one:4 -} makeItem ("one-two" :: String) >>= saveSnapshot "two"

        create ["two.html"] $ do
            route idRoute
            compile $ do
                {- two:1 -} _ <- (loadSnapshotBody "one.html" "one") :: Compiler String
                {- two:2 -} makeItem ("two-two" :: String) >>= saveSnapshot "two"
                {- two:3 -} text <- loadSnapshotBody "one.html" "two"
                {- two:4 -} makeItem (text :: String)

    ec @?= ExitSuccess

    cleanTestEnv
```

With this PR, it fails:

```
[ERROR] Hakyll.Core.Runtime.pickAndChase: Dependency cycle detected: (one.html,"two"), (two.html,"two") are inter-dependent.
```

But it passes on v4.13.4.1 (which uses the old cycle detector).

To make it easier to see what's going on, I annotated the lines of each compiler and drew the dependencies. Dotted lines are the execution sequence (they point at the next line), solid lines are dependencies (they point from `loadSnapshotBody` to `saveSnapshot`):

![deps](https://user-images.githubusercontent.com/118875/133335946-c6d67ab3-b692-4ff6-8ea5-014b96227463.png)

So there's really no cycle here.

We have to put something else in place of `depSnapshot` here, but we can't do that until we ran the current compiler until either the end (then it's "_final") or a `saveSnapshot` call.

I wish I was also inspired enough to propose a solution, but I ran out of energy already :) If you have ideas, great! If not, let this sit for a day or two and let's see if I can come up with something. (Gosh, I wish I was *this* attentive back when you sent the first PR, we'd have this sorted by now! :)

Reply:
Hmmm yes I was also bothered by this. Problem is, snapshots aren't like 'compilation stages', when you move forward through them from front to back. 

I don't think that we can construct a task graph with precise enough nodes. We know what snapshots are in the nodes downstream, but we don't know the snapshots of nodes upstream. This makes it hard (impossible?) to find dependency cycles via topological sorting like we've been doing.

I have a proposition which is simpler: instead of tracking dependencies and detecting cycles proactively, we could instead detect when we are at a fixed point. When the runtimeTodo hasn't changed after two iterations, this is an indication that there is a dependency cycle. What do you think?

Reply:
Take a look at the latest commit. I think this is much simpler and elegant. A dependency cycle is detected when *all* current tasks are 'idling' because they have requirements which are not available.